src/audio/volume/volume.c

```diff
@@ -1181,6 +1179,64 @@ static vol_zc_func vol_get_zc_function(struct comp_dev *dev)
 	return NULL;
 }
 
+/**
+ * \brief Set volume frames alignment limit.
+ * \param[in,out] source Source structure.
+ * \param[in,out] sink Sink structure.
+ */
+#if XCHAL_HAVE_HIFI3 || XCHAL_HAVE_HIFI4
+
+/**
+ * in HiFi3 or HiFi4, we use multi-way operation intrinsics, those operations
+ * require 8-byte aligned. Since there is at most SOF_IPC_MAX_CHANNELS channels,
+ * the following 2 arrays are the restriction tables of 16 and 32 bit per sample.
+ */
+
+const struct processing_alignment volume_align_16bit[SOF_IPC_MAX_CHANNELS] = {
+	{2, 2}, {4, 1}, {6, 2}, {8, 0},
+	{10, 2}, {12, 1}, {14, 2}, {16, 0}
+};
+
+const struct processing_alignment volume_align_32bit[SOF_IPC_MAX_CHANNELS] = {
+	{4, 1}, {8, 0}, {12, 1}, {16, 0},
+	{20, 1}, {24, 0}, {28, 1}, {32, 0}
+};
+
+static void volume_get_alignment(struct audio_stream *source,
+				 struct audio_stream *sink)
+{
+	uint16_t shift_idx = 0;
+	struct processing_alignment *volume_align;
+	int i;
+	uint16_t sample_size = get_sample_bytes(source->frame_fmt);
+	int frame_size = sample_size * source->channels;
+
+	if (sample_size == 2)
+		volume_align = (struct processing_alignment *)volume_align_16bit;
+	else
+		volume_align = (struct processing_alignment *)volume_align_32bit;
+
+	for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++) {
+		if (frame_size == volume_align[i].frame_size) {
```

Comment:
It looks like frame_size struct member is used only in this search loop. Is the intention that the table can be populated in not increasing channels count order? Otherwise wouldn't just `shift_idx = volume_align[source->channels].alignment_shift_idx;` work?